crypto/evp/keymgmt_lib.c

```diff
@@ -211,3 +225,259 @@ void *evp_keymgmt_util_fromdata(EVP_PKEY *target, EVP_KEYMGMT *keymgmt,
 
     return keydata;
 }
+
+int evp_keymgmt_util_has(EVP_PKEY *pk, int selection)
+{
+    size_t i, end = OSSL_NELEM(pk->pkeys);
+
+    for (i = 0; i < end && pk->pkeys[i].keymgmt != NULL; i++) {
+        const EVP_KEYMGMT *keymgmt = pk->pkeys[i].keymgmt;
+        void *keydata = pk->pkeys[i].keydata;
+
+        if (keymgmt->has != NULL)
+            return evp_keymgmt_has(keymgmt, keydata, selection);
+    }
+    /*
+     * TODO(3.0) investigate whether the lack of any |has| function should
+     * mean that the key has the components indicated by |selection| (i.e.
+     * they are assumed to be the empty set) or not.
+     * TODO(3.0) Investigate if we should return a value that indicates that
+     * this function is unsupported, leaving it to the caller to decide what
+     * to do.
+     *
+     * We currently assume that if there is no |has| function, the components
+     * indicated by |selection| are present, because that's how
+     * EVP_PKEY_missing_parameters() behaves for legacy EVP_PKEYs.
+     */
+    return 1;
+}
+
+/*
+ * Specialized function to find matching keymgmt in two EVP_PKEYs, given
+ * the return value from a |finder| function.  This function assumes that
+ * if there isn't a perfect match, the caller will have to export one of
+ * the keys to the other key's importing keymgmt, so it looks for that
+ * too.
+ *
+ * No return value.  The success is determined by the values assigned to
+ * the pointer references.
+ */
+static void find_keymgmt(EVP_KEYMGMT **keymgmt1, EVP_KEYMGMT **keymgmt2,
+                         void **keydata1, void **keydata2,
+                         EVP_PKEY *pk1, EVP_PKEY *pk2,
+                         int (*finder)(EVP_KEYMGMT *keymgmt))
+{
+    EVP_KEYMGMT *tmp_keymgmt1 = NULL, *tmp_keymgmt2 = NULL;
+    void *tmp_keydata1 = NULL, *tmp_keydata2 = NULL;
+    size_t i1, i2, end = OSSL_NELEM(pk1->pkeys);
+
+    /*
+     * Find cache elements that share the same keymgmt, which must also have
+     * function determined by |finder|.  This search is obviously O(n^2), but
+     * since the cache is fairly small, we assume that it's still cheaper
+     * than a export + import.
+     */
+    for (i1 = 0; i1 < end && pk1->pkeys[i1].keymgmt != NULL; i1++) {
+        if (finder(pk1->pkeys[i1].keymgmt)) {
+            tmp_keymgmt1 = pk1->pkeys[i1].keymgmt;
+            tmp_keydata1 = pk1->pkeys[i1].keydata;
+        }
```

Comment:
Yes.  The idea is that it may be sufficient if either of the EVP_PKEYs have a keymgmt for which `finder` returns true.
The next step will be `prepare_binary_op`, which does a cross-wise export and should end up finding the most suitable keymgmt to use.

Reply:
Probably I'm misunderstanding something here. If `finder` doesn't return true here, then when we get to the inner loop one of the following is true:
1) tmp_keymgmt1 and tmpkeydata1 are both NULL - which means the "bigbreak" condition will never be hit in that case
or
2) tmp_keymgmt1 and tmpkeydata1 have the same values as the last time we went around the outer loop. So if we didn't find a match in the inner loop last time - why go around it again?

Reply:
@levitte ?

Reply:
I believe this question needs to wait for #11074, as conditions are changing with that one.  Basically, when that one lands, any `EVP_PKEY` with `pkeys[0].keymgmt == NULL` is to be considered unassigned as far as this file goes (legacy export is taken care of in `evp_pkey_make_provided()`).